examples/greeter-extension/README.md

```diff
@@ -0,0 +1,328 @@
+`Component` are used to build the service in an extensible fashion.
+
+## Define an extension point
+
+In our scenario, we want to allow other modules to extend or customize how
+people are greeted in different languages. To achieve that, we declare the
+`greeter` extension point, which declares a contract as TypeScript interfaces
+that extensions must conform to.
+
+### Define interface for extensions
+
+An extension point interacts with unknown number of extensions. It needs to
+define one or more interfaces as contracts that each extension must implement.
+
+```ts
+/**
+ * Typically an extension point defines an interface as the contract for
+ * extensions to implement
+ */
+export interface Greeter {
+  language: string;
+  greet(name: string): string;
+}
+```
+
+### Define class for the extension point
+
+Typically an extension point is defined as a TypeScript class and bound to a
+context.
+
+```ts
+/**
+ * An extension point for greeters that can greet in different languages
+ */
+export class GreeterExtensionPoint {
+  constructor(
+    /**
+     * Inject a getter function to fetch greeters (bindings tagged with
+     * 'greeter')
+     */
+    @inject.getter(bindingTagFilter({extensionPoint: 'greeter'}))
+    private greeters: Getter<Greeter[]>,
+  ) {}
+  // ...
+}
+```
+
+To customize metadata such as `id` for the extension point, we can use
+`@extensionPoint` to decorate the class, such as:
+
+```ts
+@extensionPoint('greeter')
+export class GreeterExtensionPoint {}
+```
+
+#### Access extensions for a given extension point
+
+To simplify access to extensions for a given extension point, we use dependency
+injection to receive a `getter` function that gives us a list of greeters.
+
+```ts
+@extensionPoint('greeter')
+export class GreeterExtensionPoint {
+  constructor(
+    /**
+     * Inject a getter function to fetch greeters (bindings tagged with
+     * 'greeter')
+     */
+    @extensions() // Sugar for @inject.getter(filterByTag({extensionPoint: 'greeter'}))
+    private greeters: Getter<Greeter[]>, // ...
+  ) {}
+}
+```
+
+Please note that it's possible to add/remove greeters after the extension point
+is instantiated. The injected getter function always pick up the latest list of
+greeters. For example:
+
+```ts
+// Get the latest list of greeters
+const greeters = await this.greeters();
+```
+
+#### Implement the delegation logic
+
+Typically, the extension point implementation will get a list of registered
+extensions. For example, when a person needs to be greeted in a specific
+language, the code iterates through all greeters to find an instance that
+matches the language.
+
+```ts
+/**
+ * Greet in the given language
+ * @param language Language code
+ * @param name Name
+ */
+async greet(language: string, name: string): Promise<string> {
+  let greeting: string = '';
+  // Get the latest list of greeters
+  const greeters = await this.greeters();
+  // Find a greeter that can speak the given language
```

Comment:
If we end up with a single class implementing both greeter lookup and invocation of the `greet`, then we should at least split the implementation of this method into two parts.

```ts
export class GreeterExtensionPoint {
  // ...

async greet(language: string, name: string): Promise<string> {
  const greeter = this.findGreeterByLanguage(language);

  const greeting = greeter
    ? greeter.greet(name):
    : `Hello, ${name}`;  // Fall back to English

  if (this.options && this.options.color) {
    greeting = chalk.keyword(this.options.color)(greeting);
  }
  return greeting;
}
```

Reply:
I improved the code to have `findGreeter` and `greet`. The class is renamed to `GreetingService`.